### PR TITLE
Support OpenSSL 1.1.0

### LINKED ISSUE #2151: support OpenSSL 1.1
Fedora Rawhide has dumped OSSL 1.1 in their repository, the build log down below is the result of trying to build against it

```
/home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp: In function ‘DH* my_get_dh(int)’:
/home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:212:4: error: invalid use of incomplete type ‘DH {aka struct dh_st}’
  dh->p = BN_bin2bn(p, sp, nullptr);
    ^~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from /usr/include/openssl/comp.h:16,
                 from /usr/include/openssl/ssl.h:47,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.h:43,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:25:
/usr/include/openssl/ossl_typ.h:104:16: note: forward declaration of ‘DH {aka struct dh_st}’
 typedef struct dh_st DH;
                ^~~~~
/home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:213:4: error: invalid use of incomplete type ‘DH {aka struct dh_st}’
  dh->g = BN_bin2bn(g, sg, nullptr);
    ^~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from /usr/include/openssl/comp.h:16,
                 from /usr/include/openssl/ssl.h:47,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.h:43,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:25:
/usr/include/openssl/ossl_typ.h:104:16: note: forward declaration of ‘DH {aka struct dh_st}’
 typedef struct dh_st DH;
                ^~~~~
/home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:214:8: error: invalid use of incomplete type ‘DH {aka struct dh_st}’
  if((dh->p == nullptr) || (dh->g == nullptr))
        ^~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from /usr/include/openssl/comp.h:16,
                 from /usr/include/openssl/ssl.h:47,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.h:43,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:25:
/usr/include/openssl/ossl_typ.h:104:16: note: forward declaration of ‘DH {aka struct dh_st}’
 typedef struct dh_st DH;
                ^~~~~
/home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:214:30: error: invalid use of incomplete type ‘DH {aka struct dh_st}’
  if((dh->p == nullptr) || (dh->g == nullptr))
                              ^~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from /usr/include/openssl/comp.h:16,
                 from /usr/include/openssl/ssl.h:47,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.h:43,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:25:
/usr/include/openssl/ossl_typ.h:104:16: note: forward declaration of ‘DH {aka struct dh_st}’
 typedef struct dh_st DH;
                ^~~~~
/home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp: In member function ‘int KviSSLCertificate::fingerprintDigestId()’:
/home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:698:31: error: invalid use of incomplete type ‘X509 {aka struct x509_st}’
  int NID = OBJ_obj2nid(m_pX509->sig_alg->algorithm);
                               ^~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from /usr/include/openssl/comp.h:16,
                 from /usr/include/openssl/ssl.h:47,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.h:43,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:25:
/usr/include/openssl/ossl_typ.h:119:16: note: forward declaration of ‘X509 {aka struct x509_st}’
 typedef struct x509_st X509;
                ^~~~~~~
/home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:713:15: error: invalid use of incomplete type ‘const EVP_MD {aka const struct evp_md_st}’
  return mdType->type;
               ^~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from /usr/include/openssl/comp.h:16,
                 from /usr/include/openssl/ssl.h:47,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.h:43,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:25:
/usr/include/openssl/ossl_typ.h:91:16: note: forward declaration of ‘EVP_MD {aka struct evp_md_st}’
 typedef struct evp_md_st EVP_MD;
                ^~~~~~~~~
/home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp: In member function ‘void KviSSLCertificate::extractPubKeyInfo()’:
/home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:832:22: error: invalid use of incomplete type ‘EVP_PKEY {aka struct evp_pkey_st}’
   m_szPubKeyType = (p->type == NID_undef) ? __tr("Unknown") : OBJ_nid2ln(p->type);
                      ^~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from /usr/include/openssl/comp.h:16,
                 from /usr/include/openssl/ssl.h:47,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.h:43,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:25:
/usr/include/openssl/ossl_typ.h:93:16: note: forward declaration of ‘EVP_PKEY {aka struct evp_pkey_st}’
 typedef struct evp_pkey_st EVP_PKEY;
                ^~~~~~~~~~~
/home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:832:75: error: invalid use of incomplete type ‘EVP_PKEY {aka struct evp_pkey_st}’
 _szPubKeyType = (p->type == NID_undef) ? __tr("Unknown") : OBJ_nid2ln(p->type);
                                                                        ^~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from /usr/include/openssl/comp.h:16,
                 from /usr/include/openssl/ssl.h:47,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.h:43,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:25:
/usr/include/openssl/ossl_typ.h:93:16: note: forward declaration of ‘EVP_PKEY {aka struct evp_pkey_st}’
 typedef struct evp_pkey_st EVP_PKEY;
                ^~~~~~~~~~~
/home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp: In member function ‘void KviSSLCertificate::extractSignature()’:
/home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:855:29: error: invalid use of incomplete type ‘X509 {aka struct x509_st}’
  int i = OBJ_obj2nid(m_pX509->sig_alg->algorithm);
                             ^~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from /usr/include/openssl/comp.h:16,
                 from /usr/include/openssl/ssl.h:47,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.h:43,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:25:
/usr/include/openssl/ossl_typ.h:119:16: note: forward declaration of ‘X509 {aka struct x509_st}’
 typedef struct x509_st X509;
                ^~~~~~~
/home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:860:24: error: invalid use of incomplete type ‘X509 {aka struct x509_st}’
  for(i = 0; i < m_pX509->signature->length; i++)
                        ^~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from /usr/include/openssl/comp.h:16,
                 from /usr/include/openssl/ssl.h:47,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.h:43,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:25:
/usr/include/openssl/ossl_typ.h:119:16: note: forward declaration of ‘X509 {aka struct x509_st}’
 typedef struct x509_st X509;
                ^~~~~~~
/home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:864:50: error: invalid use of incomplete type ‘X509 {aka struct x509_st}’
   m_szSignatureContents.append(hexdigits[(m_pX509->signature->data[i] & 0xf0) >> 4]);
                                                  ^~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from /usr/include/openssl/comp.h:16,
                 from /usr/include/openssl/ssl.h:47,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.h:43,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:25:
/usr/include/openssl/ossl_typ.h:119:16: note: forward declaration of ‘X509 {aka struct x509_st}’
 typedef struct x509_st X509;
                ^~~~~~~
/home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:865:50: error: invalid use of incomplete type ‘X509 {aka struct x509_st}’
   m_szSignatureContents.append(hexdigits[(m_pX509->signature->data[i] & 0x0f)]);
                                                  ^~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from /usr/include/openssl/comp.h:16,
                 from /usr/include/openssl/ssl.h:47,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.h:43,
                 from /home/dessa/KVIrc/src/kvilib/net/KviSSL.cpp:25:
/usr/include/openssl/ossl_typ.h:119:16: note: forward declaration of ‘X509 {aka struct x509_st}’
 typedef struct x509_st X509;
                ^~~~~~~
gmake[2]: *** [src/kvilib/CMakeFiles/kvilib.dir/build.make:1599: src/kvilib/CMakeFiles/kvilib.dir/net/KviSSL.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:4125: src/kvilib/CMakeFiles/kvilib.dir/all] Error 2
gmake: *** [Makefile:128: all] Error 2
```

### PR DESCRIPTION
**Pull request by CyberShadow**
Fixes #2151.

### CODE REVIEW

File: src/kvilib/net/KviSSL.cpp [hunk 1]
```diff
@@ -828,8 +844,14 @@ void KviSSLCertificate::extractPubKeyInfo()
...
 	{
+		int type;
+#if OPENSSL_VERSION_NUMBER >= 0x10100005L
+		type = EVP_PKEY_base_id(p);
+#else
+		type = EVP_PKEY_type(p->type);
+#endif
 		m_iPubKeyBits = EVP_PKEY_bits(p);
-		m_szPubKeyType = (p->type == NID_undef) ? __tr("Unknown") : OBJ_nid2ln(p->type);
+		m_szPubKeyType = (type == NID_undef) ? __tr("Unknown") : OBJ_nid2ln(type);
```
Comment (CyberShadow) [inline src/kvilib/net/KviSSL.cpp:854]: Not sure about this one. There is no longer a way to get `p->type` - `EVP_PKEY_base_id(p)` replaces `EVP_PKEY_type(p->type)`, however I suspect that the original code was wrong and this was the intent.

**Comment by ctrlaltca** [comment]
Travis currently bundles openssl 1.0.1f, so compilation failed because new functions implemented in 1.1 are used in this PR, eg. [EVP_MD_CTX_new](https://www.openssl.org/docs/manmaster/man3/EVP_MD_CTX_new.html).

The easiest solution is probably to `#if OPENSSL_VERSION_NUMBER < 0x10100000L .. #else ... #endif` calls to new functions.

KVIrc depends on Qt that is linked to OpenSSL itself; this needs a bit of investigation, will it work if KVIrc is linked to OpenSSL 1.1 while Qt still links to a previous version?

**Comment by CyberShadow** [comment]
> Travis currently bundles openssl 1.0.1f, so compilation failed because new functions implemented in 1.1 are used in this PR, eg. EVP_MD_CTX_new.

Ah crud, those were renamed in 1.1.0. Fixed.

> KVIrc depends on Qt that is linked to OpenSSL itself; this needs a bit of investigation, will it work if KVIrc is linked to OpenSSL 1.1 while Qt still links to a previous version?

I don't know. You may want to detect that in the CMake script.

With this PR, it builds and works fine for me (on Arch Linux).

**Comment by Dessa** [comment]
see this: https://github.com/mumble-voip/mumble/commit/dff15577d6f9d62e095afc9b8772adf7f447d997
and this: https://github.com/mumble-voip/mumble/commit/f544524d4c3289bf3341d685277de0e6a6895469

**Comment by CyberShadow** [comment]
Ping

As I said above, everything works fine on Arch Linux. Any additional changes for distributions in which Qt links to an older OpenSSL would need to be looked into by someone else.

**Comment by CyberShadow** [comment]
Ping.

Is there any actionable feedback to be addressed here?

Are the Qt concerns theoretical, or is anyone actually having trouble with this patch? We shouldn't attempt to fix what we can't even reproduce.

**Comment by ctrlaltca** [comment]
I agree this one could me merged, as it:
* doesn't break any existing configuration
* adds support for a new version of a library KVIrc depends on

**Comment by CyberShadow** [comment]
Thanks!

**Pull request merged**